metals/src/main/scala/scala/meta/internal/metals/testProvider/TestSuitesIndex.scala

```diff
@@ -0,0 +1,133 @@
+package scala.meta.internal.metals.testProvider
+
+import scala.collection.concurrent.TrieMap
+import scala.collection.mutable
+
+import scala.meta.internal.metals.debug.BuildTargetClasses
+import scala.meta.internal.metals.testProvider.TestExplorerEvent._
+import scala.meta.internal.mtags
+import scala.meta.io.AbsolutePath
+
+import ch.epfl.scala.bsp4j.BuildTarget
+
+final case class FullyQualifiedName(value: String) extends AnyVal
+final case class ClassName(value: String) extends AnyVal
+
+private[testProvider] final case class SymbolsPerTarget(
+    target: BuildTarget,
+    testSymbols: TrieMap[
+      BuildTargetClasses.Symbol,
+      BuildTargetClasses.FullyQualifiedClassName
+    ]
+)
+
+private[testProvider] final case class TestFileMetadata(
+    entries: List[TestEntry],
+    hasResolvedChildren: Boolean
+)
+
+private[testProvider] final case class TestEntry(
+    buildTarget: BuildTarget,
+    path: AbsolutePath,
+    suiteName: FullyQualifiedName,
+    symbol: mtags.Symbol,
+    testClass: AddTestSuite
+)
+
+private[testProvider] final class TestSuitesIndex {
+
+  /**
+   * Cached, already discovered test suites per build target.
+   *
+   * For every test suite a.TestSuiteName Metals returns 2 symbols,
+   * one for object and one for class.
+   * However, only one of them has location in source code and it's useful.
+   * That's why FullyQualifiedClassName is used as a key.
+   * "a/TestSuiteName., a.TestSuiteName"
+   * "a/TestSuiteName#, a.TestSuiteName"
+   */
+  private val cachedTestSuites =
+    mutable.Map[
```

Comment:
I think it shouldn't be necessary since all updates are done through the `BatchedFunction` and most of the updates are also done automatically in `BatchedFunction`. The only concern is a situation when someone queries Metals exactly during `refresh` functions. 

I'll check on Metals codebase if switching implementation causes significant difference in needed time.

Reply:
Zero difference, I'll switch